### ISSUE TITLE
Mixin does not allow Generic

### ISSUE THREAD

**Issue by Novaal** [Microsoft/TypeScript#26154]
**TypeScript Version:**  3.1.0-dev.20180802

**Search Terms:**
- Typescript Mixins Generics
- Base class expressions cannot reference class type parameters.

**Code**

```ts
type Constructor<T> = new (...args: any[]) => T;

interface BasicTypes {
  user: {name: string}
}

class BasicHandler<E extends BasicTypes> {
    // some methods here
}

interface TypesA extends BasicTypes {
   user: BasicTypes["user"] & {img: string}
}

interface TypesB extends BasicTypes {
    user: BasicTypes["user"] & {email?: string}
}

function UserMixin<E extends BasicTypes, B extends Constructor<BasicHandler<E>>>(Base: B)  {
    class UserHandler extends Base { //
        constructor(...args: any[]) {
            super(...args);
        }
        setUser(user: E["user"]) {
            //...
        }
    }
    return UserHandler;
}

// error TS2562: Base class expressions cannot reference class type parameters.
class HandlerA<MergedTypes extends TypesA> extends UserMixin<MergedTypes, Constructor<BasicHandler<MergedTypes>>>(BasicHandler) {
   // some other specific methods here
}

// error TS2562: Base class expressions cannot reference class type parameters.
class HandlerB<MergedTypes extends TypesB> extends UserMixin<MergedTypes, Constructor<BasicHandler<MergedTypes>>>(BasicHandler) {
    // some other specific methods here
 }

type mergedTypes = TypesA & TypesB

const currentUser: mergedTypes["user"] = {
    name: "Peter Parker",
    img: "https://....",
    email: "pp@example.org"
}

const a: HandlerA<mergedTypes> = {} as any;
a.setUser(currentUser)

const b: HandlerB<mergedTypes> = {} as any;
b.setUser(currentUser);
```

**Expected behavior:**
Mixins should allow generic types and propagate them to the inherited class

**Actual behavior:**
Cannot set generic type to mixin

**Playground Link:** [Link](http://www.typescriptlang.org/play/#src=type%20Constructor%3CT%3E%20%3D%20new%20(...args%3A%20any%5B%5D)%20%3D%3E%20T%3B%0D%0A%0D%0Ainterface%20BasicTypes%20%7B%0D%0A%20%20user%3A%20%7Bname%3A%20string%7D%0D%0A%7D%0D%0A%0D%0Aclass%20BasicHandler%3CE%20extends%20BasicTypes%3E%20%7B%0D%0A%20%20%20%20%2F%2F%20some%20methods%20here%0D%0A%7D%0D%0A%0D%0Ainterface%20TypesA%20extends%20BasicTypes%20%7B%0D%0A%20%20%20user%3A%20BasicTypes%5B%22user%22%5D%20%26%20%7Bimg%3A%20string%7D%0D%0A%7D%0D%0A%0D%0Ainterface%20TypesB%20extends%20BasicTypes%20%7B%0D%0A%20%20%20%20user%3A%20BasicTypes%5B%22user%22%5D%20%26%20%7Bemail%3F%3A%20string%7D%0D%0A%7D%0D%0A%0D%0Afunction%20UserMixin%3CE%20extends%20BasicTypes%2C%20B%20extends%20Constructor%3CBasicHandler%3CE%3E%3E%3E(Base%3A%20B)%20%20%7B%0D%0A%20%20%20%20class%20UserHandler%20extends%20Base%20%7B%20%2F%2F%0D%0A%20%20%20%20%20%20%20%20constructor(...args%3A%20any%5B%5D)%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20super(...args)%3B%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%20%20%20%20setUser(user%3A%20E%5B%22user%22%5D)%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F...%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20return%20UserHandler%3B%0D%0A%7D%0D%0A%0D%0A%2F%2F%20error%20TS2562%3A%20Base%20class%20expressions%20cannot%20reference%20class%20type%20parameters.%0D%0Aclass%20HandlerA%3CMergedTypes%20extends%20TypesA%3E%20extends%20UserMixin%3CMergedTypes%2C%20Constructor%3CBasicHandler%3CMergedTypes%3E%3E%3E(BasicHandler)%20%7B%0D%0A%20%20%20%2F%2F%20some%20other%20specific%20methods%20here%0D%0A%7D%0D%0A%0D%0A%2F%2F%20error%20TS2562%3A%20Base%20class%20expressions%20cannot%20reference%20class%20type%20parameters.%0D%0Aclass%20HandlerB%3CMergedTypes%20extends%20TypesB%3E%20extends%20UserMixin%3CMergedTypes%2C%20Constructor%3CBasicHandler%3CMergedTypes%3E%3E%3E(BasicHandler)%20%7B%0D%0A%20%20%20%20%2F%2F%20some%20other%20specific%20methods%20here%0D%0A%20%7D%0D%0A%0D%0Atype%20mergedTypes%20%3D%20TypesA%20%26%20TypesB%0D%0A%0D%0Aconst%20currentUser%3A%20mergedTypes%5B%22user%22%5D%20%3D%20%7B%0D%0A%20%20%20%20name%3A%20%22Peter%20Parker%22%2C%0D%0A%20%20%20%20img%3A%20%22https%3A%2F%2F....%22%2C%0D%0A%20%20%20%20email%3A%20%22pp%40example.org%22%0D%0A%7D%0D%0A%0D%0Aconst%20a%3A%20HandlerA%3CmergedTypes%3E%20%3D%20%7B%7D%20as%20any%3B%0D%0Aa.setUser(currentUser)%0D%0A%0D%0Aconst%20b%3A%20HandlerB%3CmergedTypes%3E%20%3D%20%7B%7D%20as%20any%3B%0D%0Ab.setUser(currentUser)%3B)

**Related Issues:** 

- https://github.com/Microsoft/TypeScript/issues/13807
- https://github.com/Microsoft/TypeScript/issues/13979

**Comment by RyanCavanaugh** [comment]
This is intentional. See https://github.com/Microsoft/TypeScript/issues/17829 and https://github.com/Microsoft/TypeScript/pull/17922 , especially https://github.com/Microsoft/TypeScript/issues/17829#issuecomment-323020591

**Comment by rzvc** [comment]
@RyanCavanaugh, what is the reason why it is considered that the inheritance list should be outside of the scope of the class declaration? (which is the root of this problem)

It's very counter intuitive and, right now, I can't think of any reason why it should be like that.

**Comment by RyanCavanaugh** [comment]
@rzvc the type parameters of a class are not in scope in the static side of its declaration -- this should be noncontroversial as the static side of a class is a singleton and cannot vary over any type parameters (there is only one constructor function, not one class per instantiated type).

It follows that instance-specific type parameters are not in scope in the base class expression, because the base class expression is only evaluated once -- when `Foo<T> extends SomeExpr`, `SomeExpr` is only evaluated once and cannot vary over `T` because there isn't a `T` yet.

**Comment by rzvc** [comment]
@RyanCavanaugh, I'm sorry but I still don't get it. I'm not sure what you mean by the static side, as in my mind, it's all static (doesn't matter what types you feed in, you get the same code - I'm probably misunderstanding what you mean). You seem to imply that something would have to vary and I'm not sure what.

Can you please provide an explanation on the following code, on why exactly should ``T`` not be passed to ``test<T>()``, in the inheritance list of Foo, when the type checker does its thing?

```ts
function test<T>(): new () => T
{
    return null;
}

class Foo<T> extends test<T>()
{
	public constructor()
	{
		super();
	}
}
```

**Comment by RyanCavanaugh** [comment]
@rzvc let's say you wrote
```ts
new Foo<string>();
new Foo<number>();
```

How many times is `test` invoked?

The answer is 1, because `test` only gets invoked *once*, during the initialization of the class declaration.

This is different from `Foo<T>` - its constructor runs twice and can plausibly be seeing actual values of type `T`.

So it makes no sense for `string` and `number` to be provided at `test`, because there's no possible way that `test` is actually generic over `T` - its behavior cannot vary over its type parameter, because it can never deal with an actual value of `T` during its lifetime.

Put another way, let's say you wrote this:
```ts
function test<T>(argument: T): new () => T
{
    return null;
}

class Foo<T> extends test<T>(/* provide an argument */)
{
	public constructor()
	{
		super();
	}
}
```
There's nothing to "fill in the blank" with - `test` is lying about being generic; it doesn't actually vary over its argument because it *has* no arguments.

**Comment by rzvc** [comment]
Ok, I see what you mean, but it's not the whole story. What we really want to have as a generic is the return type, which is a constructor signature and gets called every time the class is instantiated.

Perhaps the type checker could reject the function call if it has generic types, that rely on the inherited one, but not if just the return type does. In the end what we care about is for the returned constructor signature to be dependent on T.

**Comment by tedchirvasiu** [comment]
What seems weird to me is that this is okay:
```typescript
class Foo<T> { }

class Bar<T> extends Foo<T> { }
```
Passing generics to anonymous classes is also okay:
```typescript
function generateFoo() {
    return class<T> { }
}

const Foo = generateFoo();

class Bar<T> extends Foo<T> { }
```
However if we try to make the anonymous class inherit from a base class received as a parameter (mixin style), it breaks:
```typescript
class Baz { }

export type Constructor<T = {}> = new (...args: any[]) => T;

function generateFoo<T extends Constructor<{}>>(Base: T) {
    return class<B> extends Base {
        constructor(...args: any[]) {
            super(...args);
        }
    }
}

const Foo = generateFoo(Baz);

class Bar<T> extends Foo<T> { }
```
Right of the bat it complains that "A mixin class must have a constructor with a single rest parameter of type 'any[]'" (which it does). It's the generic B that breaks it.

In my mind the result of generateFoo(Baz) should still be:
```typescript
class Foo<B> extends Baz { }
```

Y u do dis typscrip

**Comment by Tundon** [comment]
@RyanCavanaugh 
> let's say you wrote
> 
> ```ts
> new Foo<string>();
> new Foo<number>();
> ```
> 
> How many times is `test` invoked?
> 
> The answer is 1

 Correct me if I am wrong, but if the similar generics code is in c++, there will definitely be two `Foo` classes, and `test()` would be invoked twice (I know c++ doesn't have this mixin mechanism, just assume similar mechanism exist in c++).
Would this be where the problem come from?

**Comment by Brain2000** [comment]
We are running into the same issues trying to create very low level React mixins for specialized components.

In other languages, including C++ generics become separate "copies". I believe C++ called them templates at one point.

Not only could this solve a number of issues that we are having, but it would also allow statics to access templates, having a separate static per generic type, and perhaps other benefits as well.

**Comment by tannerntannern** [comment]
I know I'm a little late to the party, but [ts-mixer](https://www.npmjs.com/package/ts-mixer) might be able to help.  It has a decent solution to the generics problem.

**Comment by rzvc** [comment]
Hey guys, can we please reopen this issue? The problem still exists, and I believe the arguments made against it are wrong.

In reply to @RyanCavanaugh's answer:

Yes, test() gets called only once - but that's all you want anyway. The argument that it doesn't vary based on its type doesn't mean much, because generics in typescript don't really have any effect on the generated code. They're there only for the type system.

I'm going to go further and say that I was wrong too, when I partially agreed that the parameters are presenting a problem, but now I realize that there is a possible solution.

In the example you gave:
```
function test<T>(argument: T): new () => T
{
    return null;
}

class Foo<T> extends test<T>(/* provide an argument */)
{
	public constructor()
	{
		super();
	}
}
```

You could have the following:
```ts
class Foo<T> extends test<T>(Bar)
{
	public constructor()
	{
		super();
	}
}
```

What this means for Foo is that the validity of Bar is going to be checked when Foo<Whatever> will get declared. If Bar is a Whatever, then it passes, if not, it fails, so the arguments of the constructor signature simply place a constraint on T.

I understand that there are practical considerations here, because test() would get called regardless of the existence of any Foo instances, but if the type parameters on the constructor signature can be inferred or temporarily discarded (in the case where they're only used as a return type), then this should be theoretically implementable.

To summarize:
1) test() gets called only once. It doesn't vary based on its template arguments anyway.
2) The validity of test()'s parameters, can be verified independent of Foo, and they can definitely be passed before the declaration of any Foo instances, as long as their type doesn't depend on template parameters, or if it can be inferred.
3) When Foo instances are declared, test()'s parameters become constraints on what T can be.

**Comment by ShaMan123** [comment]
At the very least I would expect that declaration merging would be possible over a declared class to workaround this limitation but it isn't.
Not being able to workaround a relatively simple case as this is really 👎


```ts
// ------------------------------------ 

type Constructor<T = object, A extends any[] = any[]> = new (...a: A) => T

function FooMixin<T extends Constructor>(Base: T) { 

    class Foo extends Base { 
        foo = 'foo'
    }

    return Foo as typeof Foo & T
}

function BarMixin<T extends Constructor>(Base: T) { 

    class Bar extends FooMixin(Base as unknown as Constructor) { 
        test() { 
            console.log(this.foo)
        }
    }

    return Bar as typeof Bar & T

}

class Base<T> {
    d(x: T){
        return 2
    }
}

// ------------------------------------  reproduction

// This is where the true bug lies!
// declaration merging throws an error so it is impossible to workaround this limitation
interface A<T extends string> extends Base<T>{

}

// T & T are unrelated so it is eq to `Base<any>`
class A<T extends string> extends FooMixin(Base<T>) {

}


class NonGenericSubClass extends BarMixin(Base<string>) {

}

new A().d(4); // any :(
console.log(new A().foo)

new NonGenericSubClass().d(4);
new NonGenericSubClass().d('4');
```



<details><summary><b>Output</b></summary>

```ts
"use strict";
// ------------------------------------ 
function FooMixin(Base) {
    class Foo extends Base {
        constructor() {
            super(...arguments);
            this.foo = 'foo';
        }
    }
    return Foo;
}
function BarMixin(Base) {
    class Bar extends FooMixin(Base) {
        test() {
            console.log(this.foo);
        }
    }
    return Bar;
}
class Base {
    d(x) {
        return 2;
    }
}
// T & T are unrelated so it is eq to `Base<any>`
class A extends FooMixin((Base)) {
}
class NonGenericSubClass extends BarMixin((Base)) {
}
new A().d(4); // any :(
console.log(new A().foo);
new NonGenericSubClass().d(4);
new NonGenericSubClass().d('4');

```


</details>


<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```


</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/PTAEFpK6dv9AWAFAoC4E8AOBTUBhAewDsBnNAJwFcBjNQigHgBVQBeUQgIwCsc6ANKACCoHAA80OYgBNSoAIbEMAbQC67RcvUA+TcRwB3UAAoAdBYUAuEQEp2e5ihQAzKsToBLEqABihQgBZT3FPYhYxSWk5AhJyajoGHRMAIQVSHBtmewBvJFRkUCLQGgAbdPl-QkipWXk0jNA8lGLW0BcAzQByDsIuluKAX2dC4oocNCoKYj9O9NBMXEIXWeqAMlAnZGGCtw80bxm0imDQ8NYJWpiiMkpaegpkhszN3PyBorKK0GOa6MqAqcwql0nh5u4ANbEQiGGbzG7xe4MN4fNpScgmFGjNptGhxQilHBmUqEADmJjQAAtPKQzL1bKjWjsmSNWuNJtMfgoKIp5IscMsuTyNlsUDsUF9SPVQSw9DlUTITOIsrZ5di2uypjMAEyonbi5AgCDwE2mhCgcZYCiEGT3Q4jI3ManyGmgQyUnDjBYehbUPBcKik0ClTw4UgAQhQRpk-HKFAUBx8AFtPaSwkGqdbDPIlGIKNaeaRqp40KBXZ4k1hCFLPFxCQtqoYGBDuYR3DJva6Q0mSwn7cgwlIKC4FDQ8MIIpd-qB4um9FO6lyMrK1WKHWBWCLFF73ONylIO0Wy6XXTgAI4N0AAA2ejCUGB0V4l5SlIknUUXs+IpPnH5iVSBYgQWXZgdFyEYDWfb4ADkSAAcWkT1PBoABlKguHwF95AXGJjkA4CcEYL8f3AgoDQMYxhExMxFQAFlsABuUAjXvUArBMCV8UJYkyRMCiRGo+kRn42DiAQgwKGQtCMKw6i6MYlARPgxDJNQ9DMIqOSTC6WiukYoA)
